metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala

```diff
@@ -1,902 +1,774 @@
+  def getServiceFor(uri: String): Option[MetalsLspWorkspaceFolderService] =
+    for { // where should we log info from here ???
+      path <- uri.toAbsolutePathSafe
+      workSpaceFolder <- folderServices
+        .filter(service => path.toNIO.startsWith(service.workspace.toNIO))
+        .sortBy(_.workspace.toNIO.getNameCount())
+        .lastOption
+    } yield workSpaceFolder
+
+  def getServiceForName(
+      workspaceFolder: String
+  ): Option[MetalsLspWorkspaceFolderService] =
+    for {
+      workSpaceFolder <- folderServices
+        .find(service => service.workspaceId == workspaceFolder)
+    } yield workSpaceFolder
 
-  override def cancel(): Unit = {
-    if (isCancelled.compareAndSet(false, true)) {
-      val buildShutdown = bspSession match {
-        case Some(session) => session.shutdown()
-        case None => Future.successful(())
-      }
-      try {
-        cancelables.cancel()
-      } catch {
-        case NonFatal(_) =>
-      }
-      try buildShutdown.asJava.get(100, TimeUnit.MILLISECONDS)
-      catch {
-        case _: TimeoutException =>
-      }
+  def execForWorkspaceFolder[R](
+      uri: String,
+      exec: MetalsLspWorkspaceFolderService => CompletableFuture[R],
+      default: => R,
+  ): CompletableFuture[R] =
+    getServiceFor(uri) match {
+      case Some(service) => exec(service)
+      case None => Future.successful(default).asJava
     }
-  }
-
-  private implicit val executionContext: ExecutionContextExecutorService = ec
 
-  @volatile
-  private var userConfig: UserConfiguration = initialUserConfig
-
-  private val clientConfig = ClientConfiguration(
-    initialServerConfig,
-    initializeParams,
-  )
+  override def didOpen(
+      params: DidOpenTextDocumentParams
+  ): CompletableFuture[Unit] =
+    execForWorkspaceFolder(
+      params.getTextDocument().getUri(),
+      _.didOpen(params),
+      (),
+    )
 
-  private val languageClient = {
-    val languageClient = new ConfiguredLanguageClient(client, clientConfig)
-    // Set the language client so that we can forward log messages to the client
-    LanguageClientLogger.languageClient = Some(languageClient)
-    cancelables.add(() => languageClient.shutdown())
-    languageClient
-  }
+  override def didChange(
+      params: DidChangeTextDocumentParams
+  ): CompletableFuture[Unit] =
+    execForWorkspaceFolder(
+      params.getTextDocument().getUri(),
+      _.didChange(params),
+      (),
+    )
 
-  val statusBar: StatusBar = new StatusBar(
-    languageClient,
-    time,
-    progressTicks,
-    clientConfig,
-  )
+  override def didClose(params: DidCloseTextDocumentParams): Unit =
+    getServiceFor(params.getTextDocument().getUri()) match {
+      case Some(service) => service.didClose(params)
+      case None =>
+    }
 
-  private val embedded: Embedded = register(
-    new Embedded(
-      statusBar
+  override def didSave(
+      params: DidSaveTextDocumentParams
+  ): CompletableFuture[Unit] =
+    execForWorkspaceFolder(
```

Comment:
Maybe we could have a separate `execForWorkspaceFolder` with no default that just aliases the one with `()`